### DISCUSSION TITLE
Apple Silicon build

### DISCUSSION THREAD

**Discussion by kklobe** [diasurgical/DevilutionX#1403] [Show and tell]
If anyone's interested in an Apple Silicon build, I've got one running. It seems to be working fine; I've also been able to create and join network games with a Windows machine on the same network.

**Comment by AJenbo** [comment]
Could you post it here?

**Reply by kklobe** [reply]
CPack is not building a valid bundle under Big Sur. The codesigning is broken when fix_bundle processes the dylibs.

Forget I said anything for now - oops!

**Reply by kklobe** [reply]
OK - let's try this one.


[devilutionx-mac-arm64.dmg.zip](https://github.com/diasurgical/devilutionX/files/6274800/devilutionx-mac-arm64.dmg.zip)

**Reply by AJenbo** [reply]
Can only confirm that it doesn't run on 10.13.6 :) Hopefully we can get someone else with access to hardware to verify it.

**Reply by Bozone-Layer** [reply]
I'm running 11.3 (Beta) and I get a "damaged" error when I try to run it. I am on an M1 MacBook Pro. Is there any way I can help get this going? Sorry, no coding skills, but willing to help test builds, etc.

**Reply by kklobe** [reply]
Hey, thanks for testing this out!

Please try this one and let me know: 
[devilutionx-macos-arm64-02.dmg.zip](https://github.com/diasurgical/devilutionX/files/6292879/devilutionx-macos-arm64-02.dmg.zip)

**Reply by Bozone-Layer** [reply]
this build launched! Now I just have to find the instructions on how to install it to see if it runs. I know instructions are around here somewhere...

**Reply by AJenbo** [reply]
https://github.com/diasurgical/devilutionX#how-to-install

**Reply by Bozone-Layer** [reply]
Very much appreciated. Thank you.

The intent of my last comment was to let you know why I wasn’t able to confirm that the game actually ran. I regret that it sounded like a passive-aggressive plea for help. It was nice of you to overlook that. 

I will give it a full test today (I own the GOG version)

**Reply by Bozone-Layer** [reply]
Installed the necessary files. The game launches and plays nicely. Wow, widescreen support! This isn't just good for nostalgia. The game runs quite well on my MBP 13. Full music and voice support, and Hellfire works seamlessly. My thanks to whomever made this happen. Never thought I'd see Diablo on a Mac again; let alone running native on Apple silicon.

**Reply by AJenbo** [reply]
Thanks :)

This is also the first time that Hellfire runs natively on any Mac :D

**Reply by Bozone-Layer** [reply]
Damn, that's right. Quite a feat. I freaking LOVE open source. I'm adding this to my short list of open source Crown Jewels (Open TTD, Remnants of the Precursors, Wesnoth). If there's a donate button somewhere here, I'll find it. Thanks again for making this happen.

**Reply by kklobe** [reply]
@AJenbo are there any CMake Mac wizards around? I have been hung up for a bit here trying to get fixup_bundle (and CPack in general) to do the right thing, but it's just not working for me; I have to do the final codesign and DMG packaging manually. I have had very little luck finding info and examples (and I'm a relative CMake n00b).

**Reply by AJenbo** [reply]
> If there's a donate button somewhere here, I'll find it. 

There is not.

@kklobe sorry, don't know much about that, I don't pay for an Apple developer license so don't have access to do so either.

**Reply by kklobe** [reply]
@AJenbo ah, gotcha - who creates/tests the x86-64 Mac build? It would be great to get the arm64 variant able to build automatically from the same sources.

**Reply by AJenbo** [reply]
I did, but it's not signed by someone paying blood money to Apple so it will ask for permissions before it will be allowed to run.

**Reply by kklobe** [reply]
haha! ok, so, for an Apple Silicon binary, macOS does require a codesign, but, it can be done with the ad-hoc identity "-", no blood money required. Without a code signature, it will get the "damaged" error that Bozone was seeing with my first attempt (it will still ask for permissions with a valid codesign, which is perfectly acceptable IMHO).

What seems to happen is that fixup_bundle breaks the codesigning on the dylibs as it copies them into Frameworks. I then inserted a manual codesign command like this:

```
  if(DIST)
      install(CODE "
        include(BundleUtilities)
        fixup_bundle(${CMAKE_BINARY_DIR}/${MACOSX_BUNDLE_BUNDLE_NAME}.app \"\" \"\")
        "
        COMPONENT Runtime)
      install(CODE "execute_process(COMMAND codesign -s \"-\" ${CMAKE_BINARY_DIR}/${MACOSX_BUNDLE_BUNDLE_NAME}.app -f --deep)")
  endif()

```

which makes the app bundle all good, but then the DMG portion of CPack wipes it out and I have to manually create the image.

If you have any tips from your experience with fixup_bundle, that'd be really helpful - thanks!

**Reply by AJenbo** [reply]
Good to know, will try and look into it. 1.2.1 is coming out in just a bit over an hour.

**Comment by kklobe** [comment]
Here's a fresh arm64 build of 1.2.1, built from the release source archive.

[devilutionx-1.2.1-mac-arm64.dmg.zip](https://github.com/diasurgical/devilutionX/files/6300575/devilutionx-1.2.1-mac-arm64.dmg.zip)

**Reply by Andytizer** [reply]
Thank you this build works great. I would really appreciate the arm64 builds being incorporated into the general releases as most M1 Mac users would prefer a native ARM version, especially of the latest 1.3.0 release and future releases too.

**Reply by AJenbo** [reply]
The issue is that doing so would require a up to date Mac and the one I have access to is from 2011.

**Comment by Bozone-Layer** [comment]
I greatly appreciate the you’ve done on it. Thank you for the reply.

**Comment by mich418** [comment]
Hello everyone,

it's been a while, Diablo IV was released couple of days ago and, I got a bit nostalgic and realised I've never played first Diablo game. So... if like me you're looking for some newer build of devilutionx for Mac with Apple Silicon, here it is:

[devilutionx-1.5.0-mac-arm64.dmg.zip](https://github.com/diasurgical/devilutionX/files/11704160/devilutionx-1.5.0-mac-arm64.dmg.zip) (1.5.0-beta)

It's based on main branch (so it's 1.5.0 if I'm not mistaken), build on MacOS 13.3.1 with M2 Pro. I played it for couple of minutes before posting this message. Tried to build v1.4.1 but got some error from fmt during process (`error: static_assert failed due to requirement 'formattable' "Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt"`). The newest version builds and works fine for me.

Hope I'll fine more time to actually play it 🙂 

Have a good one!

**Reply by AJenbo** [reply]
1.5.0-beta, so won't be able to play multiplayer with release version, only other beta players. 1.5.0 proper should be out this weekend if things goes well.

**Reply by mich418** [reply]
OK, will try to keep it in mind and build it again when it's officially there 🙂

**Reply by mich418** [reply]
and here it is: 
[devilutionx-macOS-arm64.dmg.zip](https://github.com/diasurgical/devilutionX/files/11738936/devilutionx-macOS-arm64.dmg.zip)

**Reply by AJenbo** [reply]
Any idea on how to get it to build a universal binary instead?

**Reply by mich418** [reply]
I just tried with passing additional option `-DCMAKE_OSX_ARCHITECTURES="arm64;x86_64"` in command line,  but then compilation fails with some clang error:

```
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [test/liblibdevilutionx_so.dylib] Error 1
make[1]: *** [test/CMakeFiles/libdevilutionx_so.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
Undefined symbols for architecture x86_64:
  "_SDL_AllocPalette", referenced from:
      devilution::palette_init() in lto.o
...more and more ouput here
```
I'll try to investigate it later, but I'm just JS dev 🙂

**Reply by AJenbo** [reply]
ok, feel free to create an issue for this.

I was just a PHP developer when i started on this project.

**Reply by ephphatha** [reply]
I've been playing around with getting a universal build working, it's annoyingly complicated. Far as I can tell a universal binary is basically just the x86_64 binary and arm64 binary mashed into a single file with a custom header (the executable ends up with a size about the sum of each versions size), but the hard part is cross-compiling the third party libraries and making sure they're available in the app bundle.

**Reply by ephphatha** [reply]
https://github.com/orgs/Homebrew/discussions/2843 oh, looks like cross-compiling needs to avoid using brew unless we want to get really complicated with mac build scripts...

**Reply by AJenbo** [reply]
Probably doable, but does not feel exciting to have to swap out and get it working, but since we are able to do it for iOS probably most of the bits are already there.

**Reply by mich418** [reply]
thanks @ephphatha, now I realised that couple of lines before that error there were warnings about ignoring some files:
```
ld: warning: ignoring file /opt/homebrew/lib/libSDL2main.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /opt/homebrew/Cellar/sdl2_image/2.6.3_1/lib/libSDL2_image.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /opt/homebrew/Cellar/libsodium/1.0.18_1/lib/libsodium.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /opt/homebrew/lib/libfmt.10.0.0.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /opt/homebrew/lib/libSDL2.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /opt/homebrew/lib/libSDL2main.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /opt/homebrew/Cellar/sdl2_image/2.6.3_1/lib/libSDL2_image.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /opt/homebrew/Cellar/libsodium/1.0.18_1/lib/libsodium.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /opt/homebrew/lib/libfmt.10.0.0.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /opt/homebrew/lib/libSDL2.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
```
well, I can try to install homebrew for intel next to my "regular" homebrew version (like described [here](https://medium.com/mkdir-awesome/how-to-install-x86-64-homebrew-packages-on-apple-m1-macbook-54ba295230f)), and with that I can build separately x86_64 and arm64 versions, but I have no clue if I can somehow combine them later into one package, or if it's possible to tell compiler it should take packages from other directory when working on x84_64 part.

**Reply by AJenbo** [reply]
@mihauco what did you do in terms of Discord integration, just not enable it?

**Reply by mich418** [reply]
didn't enable it

**Comment by curby** [comment]
Hey folks, given that macOS is moving towards an ARM-only future, is there any hope of having a native ARM version for macOS released by the devilutionX project? I'm impressed by the large number of platforms with "official" devilutionX native binaries, and would love to see ARM macOS be supported in the same way. 

By the way, I posted here because this seemed like the most recent, active, and relevant discussion about this sort of thing but I can move it elsewhere if that would be preferable. Thanks!

**Reply by AJenbo** [reply]
It's easy to do so but as things are right now it would kill support for all Intel based Mac and doing so comes with practically no benefits.

- DevilultionX runs perfectly well via Rosetta2
- Apple is showing no indication of plans to dropping Rosetta2
- The available tools for making universal binaries are non trivial
- None of the active developers have a modern Mac

Rest assured that if tomorrow Apple decides to drop support for Intel applications we will release an ARM-only build same day, but until then we will continue hoping that someone that is a Mac native developer is willing to contribute the time to resolve the technical issues with making universal binaries.

**Reply by Trihedraf** [reply]
If you build it yourself following the [build instructions](https://github.com/diasurgical/devilutionX/blob/master/docs/building.md) it will output an Apple Silicon version instead of intel.

**Reply by curby** [reply]
> it would kill support for all Intel based Mac

To be clear, i just meant an additional version, not necessarily a Universal binary and definitely not a replacement for the current x86 binary.

> Rest assured that if tomorrow Apple decides to drop support for Intel applications we will release an ARM-only build same day

That's a great relief. I understand that developer resources (and testing hardware) are limited and it's great to know that you'll shift as needed. I appreciate all the work you do. I've been loving DevilutionX so far!

**Reply by AJenbo** [reply]
I don't see a benefit to having a separate applikation for ARM it just feels like extra work that could result in confusion for Intel users.

I would normally point out https://macsourceports.com/sourceport/devilutionx but unfortunately the have not had the time to build 1.5.0 yet. They mostly do the build manually and it's not compatible with older macOS versions.

**Comment by karbes** [comment]
I have M1 and just downloaded the source and ran commands for building arm binary, it succeeds. I don't see any reason why both Intel and ARM binaries can't be put up. If you own M1/M2 you don't want Intel binaries. Not reasonable to expect all the users to install brew + compile from the source.

**Reply by AJenbo** [reply]
> I don't see any reason why both Intel and ARM binaries can't be put up.

Ok, buy me a new ARM Mac book and it won't be an issue to do so.

The Intel version works perfectly fine on ARM, so no it is not required for everyone to install brew and compile.

**Reply by karbes** [reply]
That's the point of GitHub, community helps. Apple Silicon users can build binaries and upload.

**Reply by AJenbo** [reply]
I can't verify the content of such a build so I won't upload it as an official release.

The goal is to get universal builds working but Apple has made this painfully hard to do. As long as there are no actual issues with using Intel builds on Arm we are not going to reach for workarounds.

**Comment by karbes** [comment]
Built from the master 
[devilutionX-1.6.0-dev-macOS-arm64.dmg.zip](https://github.com/diasurgical/devilutionX/files/12922442/devilutionX-1.6.0-dev-macOS-arm64.dmg.zip)

**Reply by qndel** [reply]
it's devilutionX 🙄

**Reply by karbes** [reply]
Fixed

**Reply by Agnapot** [reply]
When I try to run this it says the file is "Damaged" with no other information.  Not sure whats wrong!  I'm on an M2 Air with 14.4.1

**Reply by TonioGela** [reply]
I can confirm it says it's damaged

**Reply by tsal** [reply]
Needs to be signed with an ad-hoc key at the very least or people will not be able to run this without disabling kernel-level security (which is generally a bad idea..)

**Reply by AJenbo** [reply]
The 1.5.3 release is a Universal build so is native on apple hardware and should also be signed with a key.

**Comment by vvaassiihhaa** [comment]
Thank you ! I have built my own apple app

0.

Install brew - go to https://brew.sh

1.

Open **Terminal**

2.

```
git clone https://github.com/diasurgical/devilutionX.git   
```

3.

```
cd devilutionX   
```

4.

```
brew bundle install
```

5.

```
cmake -S. -Bbuild -DCMAKE_BUILD_TYPE=Release -DPNG_ARM_NEON=on
cmake --build build -j $(sysctl -n hw.physicalcpu)
```

6.

copy file DIABDAT.MPQ in folder **build** 

That's it !

Video ( the old one ) https://youtu.be/xoXMroHGmbM?si=Fk25stkP6eqIZ1w_

(link removed)

**Reply by StephenCWills** [reply]
Please do not post links to unofficial downloads for game data. To get diabdat.mpq, I recommend buying the game from GOG. Blizzard now offers the game on Battle.net as well.

**Reply by vvaassiihhaa** [reply]
of course, sorry

**Reply by vvaassiihhaa** [reply]
I didn't know.

In the evening, I will try to figure out the version of my clone.

How can I do that ?

**Reply by AJenbo** [reply]
- Lower left in the menu
- Check properties just like any other app
- press v in game
- run it in a terminal with --version

**Reply by AJenbo** [reply]
It's probably a build of the 1.6.0 alpha currently in development if you did not specify it

**Reply by vvaassiihhaa** [reply]
yes, it is DevilutionX v1.6.0-dev-Release-322c627f5

and it's work perfect ! :)

apple M2 

you can see my first try to play here https://www.twitch.tv/videos/2054116538

**Reply by Agnapot** [reply]
When trying to build this I get the following cmake error when trying to find libsodium.  

CMake Error at /opt/homebrew/Cellar/cmake/3.29.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find sodium (missing: sodium_LIBRARY_RELEASE
  sodium_LIBRARY_DEBUG) (found version "1.0.19")

Not sure why, I seem to have libsodium installed, but I'm not sure how to check what versions I have, or why it might not be finding it.  Any assistance is appreciated!

**Reply by StephenCWills** [reply]
I don't know much about brew, but the libraries would probably be somewhere under `/opt/homebriew/Cellar` with the `.so` extension. You might be able to work around the issue by adding `-DDEVILUTIONX_SYSTEM_LIBSODIUM=OFF` to the first `cmake` command.

**Reply by Agnapot** [reply]
That worked!  Thank you!  Not sure what the catch was, but it found some version of libsodium that seemed suitable. This is what the cmake said this time:
-- 📚 libsodium: dynamic library from source
It hung there a bit, then moved on, so, success!

**Reply by AJenbo** [reply]
When you turn off using the system installed package it will download the dependency from source and build it, which is why it hung for a bit.

**Comment by cchisholm-needs-help-coding** [comment]
I have an M1 macbook.  I've been trying to build from source, but currently stuck on an issue.  I have successfully compiled and get the package, but get an error "can't open diabat.mpq or spawn.mpq" when I try to run it.  Has anyone run into that?  It's in the same folder as the package (DevolutionX), I tried both upper and lower case.  Seems like I need to specify a working directory somewhere or something.

One thing to note for others- I had previously been running v1.21 released in this thread, and the diablo.ini was incompatible, so make sure if you've been running an older version and compile a newer version to backup/delete diablo.ini in ~/Library/Application Support/diasurgical/devilution.  You'll get some kind of error like 'error line 1: ....'

**Reply by Trihedraf** [reply]
Place mpq files in ~/Library/Application Support/diasurgical/devilution.

**Reply by cchisholm-needs-help-coding** [reply]
that worked!!  thanks

**Reply by AJenbo** [reply]
The 1.5.4 release has native Apple support so you don't actually need to build it your self for that.
